### PR TITLE
Windows: enable tests and envoy-static.exe pdb file

### PR DESCRIPTION
**Pull request by sunjayBhatia**
Commit Message: Windows: enable tests and envoy-static.exe pdb file
Additional Description:
- Introduce debugging output for msvc-cl and clang-cl opt binaries in
      order to be able to unwind stack traces from windows core files.
- Mark additional tests fails on windows based on clang-cl build
      observed test failures
- Work around problematic googletests with clang-cl compilation
- Solve problematic missing override declarations of mocks by temporarilly
adding -Wno-inconsistent-missing-override for clang. (It appears
googletest on clang-cl isn't handling something that our gcc, clang and
msvc-cl builds all ignore without the warning override.
- Workaround THROW tests which googletest with clang-cl are framing as
as -Wdiscard errors.
- Workaround nodiscard next() method error
- Follow up about discarded results of singleton construction and other
discarded odd or apparently irrelevant function call results.
- Holding on adding clang-cl CI until LLVM 11.0.0 can be used on Windows
  with Bazel (requires Bazel 3.7.1 and 4.0.0 releases)

Risk Level: Low (only affects Windows build)
Testing: Verified locally
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A
Related to #11974 (partially addresses)

### CODE REVIEW

**Comment by sunjayBhatia** [comment]
cc @envoyproxy/windows-dev

File: test/common/http/http2/BUILD
Comment (wrowe) [inline test/common/http/http2/BUILD:58]: This is what I'm looking for some possible assistance from google folks, @htuch - we are having serious problems deciphering why clang-cl is failing in a manner unique to building opt (/O2) on Windows using clang-cl, and in no other environments or compilers.  This particular link failure reads as follows;
```
ERROR: C:/users/wrowe/dev/envoy/test/common/http/http2/BUILD:40:14: Linking of rule '//test/common/http/http2:codec_impl_test' failed (Exit 1)
lld-link: error: undefined symbol: public: virtual __cdecl Envoy::Http::ConnectionCallbacks::~ConnectionCallbacks(void)
>>> referenced by bazel-out/x64_windows-opt/bin/test/common/http/http2/_objs/codec_impl_test/codec_impl_test.obj:(int `public: __cdecl Envoy::Http::Http2::Http2CodecImplTestFixture::Http2CodecImplTestFixture(class std::tuple<unsigned int, unsigned int, unsigned int, unsigned int>, class std::tuple<unsigned int, unsigned int, unsigned int, unsigned int>)'::`1'::dtor$69)
>>> referenced by bazel-out/x64_windows-opt/bin/test/common/http/http2/_objs/codec_impl_test/codec_impl_test.obj:(int `public: __cdecl Envoy::Http::Http2::Http2CodecImplTestFixture::Http2CodecImplTestFixture(class std::tuple<unsigned int, unsigned int, unsigned int, unsigned int>, class std::tuple<unsigned int, unsigned int, unsigned int, unsigned int>)'::`1'::dtor$76)
>>> referenced by bazel-out/x64_windows-opt/bin/test/common/http/http2/_objs/codec_impl_test/codec_impl_test.obj:(public: void __cdecl Envoy::Http::MockServerConnectionCallbacks::`vbase dtor'(void))
>>> referenced by bazel-out/x64_windows-opt/bin/test/common/http/http2/_objs/codec_impl_test/codec_impl_test.obj:(public: void __cdecl Envoy::Http::MockConnectionCallbacks::`vbase dtor'(void))>>> referenced by bazel-out/x64_windows-opt/bin/test/common/http/http2/_objs/codec_impl_test/codec_impl_test.obj:(public: virtual __cdecl Envoy::Http::Http2::Http2CodecImplTestFixture::~Http2CodecImplTestFixture(void))
>>> referenced by bazel-out/x64_windows-opt/bin/test/common/http/http2/_objs/codec_impl_test/codec_impl_test.obj:(public: virtual __cdecl Envoy::Http::Http2::Http2CodecImplTestFixture::~Http2CodecImplTestFixture(void))
>>> referenced by bazel-out/x64_windows-opt/bin/test/common/http/http2/_objs/codec_impl_test/codec_impl_test.obj:(int `public: __cdecl Envoy::Http::Http2::Http2CodecImplTestFixture::Http2CodecImplTestFixture(void)'::`1'::dtor$18)
>>> referenced by bazel-out/x64_windows-opt/bin/test/common/http/http2/_objs/codec_impl_test/codec_impl_test.obj:(int `public: __cdecl Envoy::Http::Http2::Http2CodecImplTestFixture::Http2CodecImplTestFixture(void)'::`1'::dtor$25)
Target //test/common/http/http2:codec_impl_test failed to build
```
Any chance you can find some support for us researching what might be happening here?

**Review by wrowe — COMMENTED** [review]
Reassigning to @htuch since https://github.com/envoyproxy/envoy/issues/11974 was instigated by himself.

Note this patch includes one commit of miscellaneous failure resolutions, which will actually land in their own PR. As I'm converting this to ready for review, we are done force-pushing changes, but we will merge master once that other straightforward PR is done and these will be cleaned up here.

Also, some 16 tests fail to compile. These are set up to all deliberately fail to build in the clang-cl pipeline after-the-fact, but the failures can be examined in that build pipeline's log. It's possible to merge this PR and simply skip those tests in the short-term, but if we can identify a fix in the coming week, we may as well hold off introducing a new pipeline till that's solved.

**Comment by wrowe** [comment]
Currently stalled to pick up clang-cl 11 and reproduce the linkage error.  See https://github.com/bazelbuild/rules_cc/pull/88

File: test/common/http/http2/BUILD
Comment (wrowe) [reply test/common/http/http2/BUILD:58]: Quick update for @htuch and @lizan - we attempted to verify this error with the llvm 11.0.0 release of clang-cl, however that update breaks bazel. Holding for this 1-line PR to be accepted to allow us to pick up the new clang-cl, and confirm whether the bug remains or has been resolved.

https://github.com/bazelbuild/rules_cc/pull/88

/wait

File: test/common/config/datasource_test.cc
Comment (htuch) [reply test/common/config/datasource_test.cc:486]: Definitely would rather not have this void cast. I'll follow up on this offline, but one option is to build the cast explicitly into the `EXPECT_THROW_WITH_MESSAGE` macro, which is something we define.

Comment (htuch) [reply test/common/config/datasource_test.cc:486]: LMK if this is still an issue on Clang 11.

**Comment by wrowe** [comment]
rules_cc now addressed, awaiting https://github.com/bazelbuild/bazel/pull/12458 and subsequent bazel release. Still /wait ing

**Comment by wrowe** [comment]
> awaiting bazelbuild/bazel #12458 and subsequent bazel release

Ready when Bazel 4.0.0 is released to bump envoy-bazel-build toolchain.

File: test/common/http/http2/BUILD
Comment (wrowe) [reply test/common/http/http2/BUILD:58]: This does appear to be resolved in clang 11.0.0 release.

File: test/common/config/datasource_test.cc
Comment (wrowe) [reply test/common/config/datasource_test.cc:486]: This is still an issue with -Wall -Werror; however, it makes sense to keep this potential defect detection in source/... --- but it doesn't seem interesting to track discarded results in test/... sources. Moved the warning suppression to the appropriate tests-only copt logic for clang-cl builds, and left a handful of corrections for source/... --- including documentation of why the results are discarded.

**Comment by lizan** [comment]
@wrowe we can bump rules_cc without bump bazel version, will just update https://github.com/envoyproxy/envoy/blob/master/bazel/repository_locations.bzl#L535 work?

**Comment by wrowe** [comment]
> @wrowe we can bump rules_cc without bump bazel version, will just update https://github.com/envoyproxy/envoy/blob/master/bazel/repository_locations.bzl#L535 work?

Unfortunately combining the new rules_cc for envoy with 3.6.0 bazel always results in these two exceptions... building for msvc-cl;
```
ERROR: C:/_eb/external/local_config_cc/BUILD:501:24: unterminated string literal at eol
ERROR: C:/_eb/external/local_config_cc/BUILD:502:1: syntax error at 'Target': expected ,
ERROR: C:/_eb/external/local_config_cc/BUILD:502:17: unterminated string literal at eol
ERROR: C:/_eb/external/local_config_cc/BUILD:503:20: unterminated string literal at eol
ERROR: C:/_eb/external/local_config_cc/BUILD:504:22: unterminated string literal at eol
ERROR: C:/_eb/external/local_config_cc/BUILD:517:9: unterminated string literal at eol
ERROR: C:/_eb/external/local_config_cc/BUILD:518:17: unterminated string literal at eol
ERROR: C:/_eb/external/bazel_tools/tools/cpp/BUILD:469:14: Target '@bazel_tools//tools/cpp:compiler' depends on toolchain '@local_config_cc//:cc-compiler-x64_windows', which cannot be found: no such target '@local_config_cc//:cc-compiler-x64_windows': target 'cc-compile
r-x64_windows' not declared in package '' defined by C:/_eb/external/local_config_cc/BUILD'
ERROR: While resolving configuration keys for //test/common/access_log:access_log_manager_impl_test: Analysis failed
ERROR: Analysis of target '//test/common/access_log:access_log_manager_impl_test' failed; build aborted: Analysis failed
```
and when building for clang-cl;
```
Starting local Bazel server and connecting to it...
ERROR: C:/_ec/external/local_config_cc/BUILD:209:20: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:210:1: syntax error at 'Target': expected ,
ERROR: C:/_ec/external/local_config_cc/BUILD:210:22: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:223:9: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:224:17: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:285:20: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:286:22: syntax error at ';': expected expression
ERROR: C:/_ec/external/local_config_cc/BUILD:287:22: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:300:9: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:301:17: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:362:20: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:363:22: syntax error at ';': expected expression
ERROR: C:/_ec/external/local_config_cc/BUILD:364:22: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:377:9: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:378:17: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:439:20: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:440:22: syntax error at ';': expected expression
ERROR: C:/_ec/external/local_config_cc/BUILD:441:22: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:454:9: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:455:17: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:516:24: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:517:17: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:518:20: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:519:22: syntax error at ';': expected expression
ERROR: C:/_ec/external/local_config_cc/BUILD:520:22: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:533:9: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:534:17: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:535:9: unterminated string literal at eol
ERROR: C:/_ec/external/local_config_cc/BUILD:536:17: unterminated string literal at eol
ERROR: C:/_ec/external/bazel_tools/tools/cpp/BUILD:469:14: Target '@bazel_tools//tools/cpp:compiler' depends on toolchain '@local_config_cc//:cc-compiler-x64_windows', which cannot be found: no such target '@local_config_cc//:cc-compiler-x64_windows': target 'cc-compile
r-x64_windows' not declared in package '' defined by C:/_ec/external/local_config_cc/BUILD'
ERROR: While resolving configuration keys for //test/common/access_log:access_log_manager_impl_test: Analysis failed
ERROR: Analysis of target '//test/common/access_log:access_log_manager_impl_test' failed; build aborted: Analysis failed
```

**Comment by lizan** [comment]
Hmm interesting, would be good to see whether 3.7 fixes this or not if we want to be unblocked.

**Comment by wrowe** [comment]
> Hmm interesting, would be good to see whether 3.7 fixes this

Indeed, it is interesting, https://github.com/bazelbuild/bazel/issues/12188#issuecomment-729773418 should tell us whether 3.7.1 resolves it. Evaluating rc2 at the moment.

**Comment by wrowe** [comment]
Apologies, had to force-push the DCO I'd missed in the first push.

**Comment by wrowe** [comment]
This is, finally, ready for final review and adoption ahead of integrating LLVM clang-cl in 11.0.0 (https://github.com/envoyproxy/envoy/pull/14135).

We will need to watch in the coming days and mark any newfound exceptions promptly as "flaky_on_windows", but there are future patches in the pipeline to automate this. I'd only ask that we note which specific test(s) fail (or timeout) to make quick progress on diagnosing the last few flakes.

File: .bazelrc
Comment (htuch) [inline .bazelrc:330]: Is the plan to do this in this PR?

File: bazel/envoy_internal.bzl
Comment (htuch) [reply bazel/envoy_internal.bzl:55]: Can you add this to the .bzl comments so this intuition is kept for posterity?

File: test/common/network/socket_option_test.h
Comment (htuch) [inline test/common/network/socket_option_test.h:69]: Not a huge fan of this one; I think we had some thread around trying to get to the bottom of inconsistency here; who has the potato?

File: test/extensions/transport_sockets/tls/ssl_socket_test.cc
Comment (htuch) [inline test/extensions/transport_sockets/tls/ssl_socket_test.cc:401]: I think it would be better to split the test and build changes where possible, upstream these tests changes in a dedicated PR.

File: test/test_common/environment.cc
Comment (htuch) [inline test/test_common/environment.cc:345]: I'm a little worried about this proliferating in options; should we have a `struct` for options here? We can keep the existing signature for backwards compatibility as a wrapper to avoid changing the entire world.

File: test/extensions/transport_sockets/tls/ssl_socket_test.cc
Comment (wrowe) [reply test/extensions/transport_sockets/tls/ssl_socket_test.cc:401]: There are a ridiculously small number of source changes here, which are all easily understood from the primary commit context. If you insist, I'll split it.

**Review by wrowe — APPROVED** [review]
This appears complete to me. @lizan has identified issues with our [coverage tests](https://github.com/envoyproxy/envoy/pull/14123#issuecomment-731709442) which suggest this will not be workable as bazel 3.7.1 lands, so there is less urgency and time for further discussion, @htuch.

File: test/test_common/environment.cc
Comment (wrowe) [reply test/test_common/environment.cc:345]: I can't envision additional flags, so a struct seems like overkill.

File: .bazelrc
Comment (htuch) [reply .bazelrc:330]: Can you file a tracking issue?

File: bazel/envoy_internal.bzl
Comment (htuch) [reply bazel/envoy_internal.bzl:55]: You might be technically correct here, the best kind of correct. I think this is not really as helpful as something like "do not collect debuginfo for the hundreds of test executables. (We would build fastbuild or dbg individually to troubleshoot a crashing test.)"

File: test/extensions/transport_sockets/tls/ssl_socket_test.cc
Comment (htuch) [reply test/extensions/transport_sockets/tls/ssl_socket_test.cc:401]: I disagree with the definition of ridiculously small ;-) I'm not going to insist, but as a reviewer I think it makes life easier to split these concerns usually.

File: test/test_common/environment.cc
Comment (htuch) [reply test/test_common/environment.cc:345]: Arguably it makes the tests that have been modified easier to read (should the default be true or false?), but I don't feel that strongly.

File: test/common/network/cidr_range_test.cc
Comment (htuch) [inline test/common/network/cidr_range_test.cc:71]: Why was this there before? Was it testing to ensure that an exception wasn't thrown? Or just an accidental inclusion?

File: bazel/envoy_internal.bzl
Comment (wrowe) [reply bazel/envoy_internal.bzl:56]: @htuch have a look at the new iteration and let me know if that's helpful.

File: source/extensions/clusters/redis/redis_cluster.cc
Comment (wrowe) [inline source/extensions/clusters/redis/redis_cluster.cc:279]: @htuch - this is the only non-build, non-test change in this entire patchset. Would you like me to break it out for you as an independently reviewed PR, or are we ok to merge and iterate at this point? It's the remaining example of a discarded no-op result in the //source/... since I re-merged.

File: test/common/network/cidr_range_test.cc
Comment (wrowe) [reply test/common/network/cidr_range_test.cc:71]: There's a google test macro for that, so I'd presume it was laziness after development, perhaps?

File: test/extensions/transport_sockets/tls/ssl_socket_test.cc
Comment (wrowe) [reply test/extensions/transport_sockets/tls/ssl_socket_test.cc:401]: I've commented above on the one //source/... case we should consider peeling away from this patch, everything else is test or build.

Comment (wrowe) [reply test/extensions/transport_sockets/tls/ssl_socket_test.cc:401]: There are no build changes that don't depend on the test changes, and vis-a-versa. So there would be no means to pass CI, so I'm really unclear on your ask.

File: source/extensions/clusters/redis/redis_cluster.cc
Comment (htuch) [reply source/extensions/clusters/redis/redis_cluster.cc:279]: No, I was not asking for this. I was pointing out that the test changes alone could have been upstreamed independently. Anyway, NBD.

File: test/test_common/environment.cc
Comment (wrowe) [reply test/test_common/environment.cc:345]: I'm personally a fan of bitwise flags, and was annoyed as we wrote this that named args aren't a c++ thing, so I'll rewrite it with enums if you insist. But I'd like a use case that suggests a third toggle before we bother.

**Review by htuch — APPROVED** [review]

**Review by htuch — APPROVED** [review]

**Pull request merged**
